packages/core/core/src/PackagerRunner.js

```diff
@@ -430,58 +423,164 @@ export default class PackagerRunner {
+            str.match(/@@HASH_REFERENCE_\w{8}/g) ?? [],
+          );
+          size += buf.length;
+          hash.update(buf);
+          boundaryStr = str.slice(str.length - REF_LENGTH - 1);
+        }),
+      ),
+    );
 
     if (map != null) {
-      let mapKey = getMapKey(cacheKey);
-      await this.options.cache.setStream(mapKey, blobToStream(map));
+      await this.options.cache.setStream(cacheKeys.map, blobToStream(map));
     }
+
+    let info = {size, hash: hash.digest('hex'), hashReferences};
+    await this.options.cache.set(cacheKeys.info, info);
+    return info;
   }
 }
 
 function writeFileStream(
   fs: FileSystem,
   filePath: FilePath,
   stream: Readable,
+  hashReferences: Array<string>,
+  hashRefToNameHash: Map<string, string>,
   options: ?FileOptions,
 ): Promise<number> {
   return new Promise((resolve, reject) => {
+    let initialStream = hashReferences.length
+      ? stream.pipe(replaceStream(hashRefToNameHash))
+      : stream;
     let fsStream = fs.createWriteStream(filePath, options);
-    stream
+    initialStream
       .pipe(fsStream)
       // $FlowFixMe
       .on('finish', () => resolve(fsStream.bytesWritten))
       .on('error', reject);
   });
 }
 
+function replaceStream(hashRefToNameHash) {
+  let boundaryStr = '';
+  return new Transform({
+    transform(chunk, encoding, cb) {
+      let str = boundaryStr + chunk.toString();
+      let replaced = str.replace(/@@HASH_REFERENCE_\w{8}/g, match => {
+        return hashRefToNameHash.get(match) || match;
+      });
+      boundaryStr = replaced.slice(replaced.length - REF_LENGTH - 1);
+      let strUpToBoundary = replaced.slice(0, replaced.length - REF_LENGTH - 1);
+      cb(null, strUpToBoundary);
+    },
+
+    flush(cb) {
+      cb(null, boundaryStr);
+    },
+  });
+}
+
 function getContentKey(cacheKey: string) {
   return md5FromString(`${cacheKey}:content`);
 }
 
 function getMapKey(cacheKey: string) {
   return md5FromString(`${cacheKey}:map`);
 }
+
+function getInfoKey(cacheKey: string) {
+  return md5FromString(`${cacheKey}:info`);
+}
+
+function assignComplexNameHashes(
+  hashRefToNameHash,
+  bundles,
+  bundleInfoMap,
+  hashRefToId,
+) {
+  for (let bundle of bundles) {
+    if (hashRefToNameHash.get(bundle.hashReference) != null) {
+      continue;
+    }
+
+    let includedBundles = getBundlesIncludedInHash(
+      bundle.id,
+      bundleInfoMap,
+      hashRefToId,
+    );
+
+    hashRefToNameHash.set(
+      bundle.hashReference,
+      md5FromString(
+        [...includedBundles]
+          .map(bundleId => bundleInfoMap[bundleId].hash)
+          .join(':'),
+      ).slice(-8),
+    );
+  }
+}
+
+function getBundlesIncludedInHash(
+  bundleId,
+  bundleInfoMap,
+  hashRefToId,
+  included = new Set(),
+) {
+  included.add(bundleId);
+  for (let hashRef of bundleInfoMap[bundleId].hashReferences) {
+    let referencedId = hashRefToId[hashRef];
+    if (!included.has(referencedId)) {
+      for (let ref in getBundlesIncludedInHash(
```

Comment:
nit: `for (let ref of Object.keys(getBundlesIncludedInHash(...`

Reply:
woops, this should be a `for...of` loop

Reply:
It doesn't really need to be a `Set` anymore since I'm checking to see if the referenced id is already in the collection before recursing. I'll change it to an array and instead do `included.push(...getBundlesIncludedInHash())`